### ISSUE TITLE
UDP Proxy Protocol not functioning

### ISSUE THREAD

**Issue by erroltuparker** [mholt/caddy-l4#269]
Hi All,

It appears that there is some issue with the proxy protocol implemented for UDP as my server recieves this response when trying to proxy.

```
[2024-11-18 00:00:04 Local] [192.168.200.1:37825] [UDPPROXY] System.ArgumentNullException: Value cannot be null. (Parameter 'address')
   at System.ArgumentNullException.Throw(String paramName)
   at DnsServerCore.Dns.DnsServer.ReadUdpRequestAsync(Socket udpListener, DnsTransportProtocol protocol)
[2024-11-18 00:00:04 Local] [192.168.200.1:37825] [UDPPROXY] System.IO.InvalidDataException: The stream does not contain PROXY protocol header.
   at TechnitiumLibrary.Net.ProxyProtocol.ProxyProtocolStream.CreateAsServerAsync(Stream baseStream, CancellationToken cancellationToken) in Z:\Technitium\Projects\TechnitiumLibrary\TechnitiumLibrary.Net\ProxyProtocol\ProxyProtocolStream.cs:line 138
   at DnsServerCore.Dns.DnsServer.ReadUdpRequestAsync(Socket udpListener, DnsTransportProtocol protocol) in Z:\Technitium\Projects\DnsServer\DnsServerCore\Dns\DnsServer.cs:line 409
```

But when I use nginx, it works fine. I did raise this with the provider and they also confirmed my findings

https://github.com/TechnitiumSoftware/DnsServer/issues/1107

TCP works fine for proxying as I'm also utilizing that, but I would like to cover both sides of it.

The thing I see on the caddy side when this is run is the below errors.

```
{"level":"debug","ts":1731887445.6253898,"logger":"layer4","msg":"matching","remote":"172.253.247.61:39349","error":"consumed all prefetched bytes","matcher":"layer4.matchers.dns","matched":false}
{"level":"debug","ts":1731887445.6254132,"logger":"layer4","msg":"prefetched","remote":"172.253.247.61:39349","bytes":48}
{"level":"debug","ts":1731887445.6254213,"logger":"layer4","msg":"matching","remote":"172.253.247.61:39349","matcher":"layer4.matchers.dns","matched":true}
{"level":"debug","ts":1731887446.644327,"logger":"layer4","msg":"matching","remote":"172.68.209.70:19458","error":"consumed all prefetched bytes","matcher":"layer4.matchers.dns","matched":false}
{"level":"debug","ts":1731887446.64437,"logger":"layer4","msg":"prefetched","remote":"172.68.209.70:19458","bytes":48}
{"level":"debug","ts":1731887446.6443822,"logger":"layer4","msg":"matching","remote":"172.68.209.70:19458","matcher":"layer4.matchers.dns","matched":true}
{"level":"debug","ts":1731887446.6450343,"logger":"layer4","msg":"matching","remote":"172.68.209.70:19483","error":"consumed all prefetched bytes","matcher":"layer4.matchers.dns","matched":false}
{"level":"debug","ts":1731887446.6450565,"logger":"layer4","msg":"prefetched","remote":"172.68.209.70:19483","bytes":52}
{"level":"debug","ts":1731887446.6450644,"logger":"layer4","msg":"matching","remote":"172.68.209.70:19483","matcher":"layer4.matchers.dns","matched":true}
```

This is what I have set up for the block
```
	layer4 {
		tcp/:53 {
			@tdns dns
			route @tdns {
				proxy {
					upstream tcp/ext-dns:538
					proxy_protocol v2
				}
			}
		}
		udp/:53 {
			@udns dns
			route @udns {
				proxy {
					proxy_protocol v2
					upstream {
						dial udp/ext-dns:538
					}
				}
			}
		}
}
```

And this is what I set up with nginx to confirm that it was its definitely an issue with caddy-l4
```
stream {
    log_format proxy '$remote_addr [$time_local] '
                 '$protocol $status $bytes_sent $bytes_received '
                 '$session_time "$upstream_addr" '
                 '"$upstream_bytes_sent" "$upstream_bytes_received" "$upstream_connect_time"';
    upstream dns {
        server ext-dns:538;
    }

    server {
        listen 53 udp;
        proxy_pass dns;
        proxy_protocol on;
    }

    access_log /config/log/nginx/access-dns.log proxy buffer=32k;


}
```

**Comment by mohammed90** [comment]
cc/ @WeidiDeng

**Comment by ydylla** [comment]
I think the layer4 proxy handler just does not intend to support proxy protocol with udp.  
I am not sure how proxy protocol over udp is supposed to work, but it probably has to send the header in front of each packet instead of only once.

**Comment by lxhao61** [comment]
Any plans to fix this BUG?

**Comment by WeidiDeng** [comment]
@erroltuparker can you try `xcaddy build --with github.com/mholt/caddy-l4=github.com/WeidiDeng/caddy-l4@udp-pp` to see if it's fixed?

**Comment by erroltuparker** [comment]
hi @WeidiDeng, i just tried it and still the same errors

Application endpoint

```
[2025-01-12 22:12:42 UTC] [192.168.86.5:41931] [UDPPROXY] System.IO.InvalidDataException: The stream does not contain PROXY protocol header.
   at TechnitiumLibrary.Net.ProxyProtocol.ProxyProtocolStream.CreateAsServerAsync(Stream baseStream, CancellationToken cancellationToken) in Z:\Technitium\Projects\TechnitiumLibrary\TechnitiumLibrary.Net\ProxyProtocol\ProxyProtocolStream.cs:line 138
   at DnsServerCore.Dns.DnsServer.ReadUdpRequestAsync(Socket udpListener, DnsTransportProtocol protocol) in Z:\Technitium\Projects\DnsServer\DnsServerCore\Dns\DnsServer.cs:line 415
[2025-01-12 22:12:42 UTC] [192.168.86.5:41931] [UDPPROXY] System.FormatException: An invalid IP address was specified.
 ---> System.Net.Sockets.SocketException (22): Invalid argument
   --- End of inner exception stack trace ---
   at TechnitiumLibrary.Net.ProxyProtocol.ProxyProtocolStream.ParseVersion1(String value) in Z:\Technitium\Projects\TechnitiumLibrary\TechnitiumLibrary.Net\ProxyProtocol\ProxyProtocolStream.cs:line 245
   at TechnitiumLibrary.Net.ProxyProtocol.ProxyProtocolStream.CreateAsServerAsync(Stream baseStream, CancellationToken cancellationToken)
   at DnsServerCore.Dns.DnsServer.ReadUdpRequestAsync(Socket udpListener, DnsTransportProtocol protocol) in Z:\Technitium\Projects\DnsServer\DnsServerCore\Dns\DnsServer.cs:line 415
```

Caddy Layer4

```
{"level":"debug","ts":1736719962.905754,"logger":"layer4","msg":"matching","remote":"206.189.140.177:48228","error":"consumed all prefetched bytes","matcher":"layer4.matchers.dns","matched":false}
{"level":"debug","ts":1736719962.905816,"logger":"layer4","msg":"prefetched","remote":"206.189.140.177:48228","bytes":39}
{"level":"debug","ts":1736719962.9058948,"logger":"layer4","msg":"matching","remote":"206.189.140.177:48228","matcher":"layer4.matchers.dns","matched":true}
{"level":"debug","ts":1736719982.9061859,"logger":"layer4","msg":"matching","remote":"206.189.140.177:34585","error":"consumed all prefetched bytes","matcher":"layer4.matchers.dns","matched":false}
{"level":"debug","ts":1736719982.9062274,"logger":"layer4","msg":"prefetched","remote":"206.189.140.177:34585","bytes":28}
{"level":"debug","ts":1736719982.9062378,"logger":"layer4","msg":"matching","remote":"206.189.140.177:34585","matcher":"layer4.matchers.dns","matched":true}
{"level":"debug","ts":1736720002.9057178,"logger":"layer4","msg":"connection stats","remote":"206.189.140.177:48228","read":78,"written":0,"duration":39.999970735}
{"level":"debug","ts":1736720022.9075844,"logger":"layer4","msg":"connection stats","remote":"206.189.140.177:34585","read":56,"written":0,"duration":40.001406982}
```

**Comment by WeidiDeng** [comment]
@erroltuparker How do you deploy [DnsServer](https://github.com/TechnitiumSoftware/DnsServer) using docker? I'll need to test it locally to figure out what's wrong.

**Comment by lxhao61** [comment]
@WeidiDeng Can the following example be used for testing?
```
{
  "apps": {
    "layer4": {
      "servers": {
        "tcpsni": {
          "listen": [":443"],
          "routes": [{
            "match": [{
              "tls": {}
            }],
            "handle": [{
              "handler": "proxy",
              "upstreams": [{
                "dial": ["127.0.0.1:8443"]
              }],
              "proxy_protocol": "v2"
            }]
          }]
        },
        "udpsni": {
          "listen": ["udp/:443"],
          "routes": [{
            "handle": [{
              "handler": "proxy",
              "upstreams": [{
                "dial": ["udp/127.0.0.1:8443"]
              }],
              "proxy_protocol": "v2"
            }]
          }]
        }
      }
    },
    "http": {
      "servers": {
        "srvh3": {
          "listen": ["127.0.0.1:8443"],
          "listener_wrappers": [{
            "wrapper": "proxy_protocol",
            "allow": ["127.0.0.1/32"]
          },
          {
            "wrapper": "tls"
          }],
          "routes": [{
            "handle": [{
              "handler": "headers",
              "response": {
                "set": {
                  "Alt-Svc": ["h3=\":443\"; ma=2592000"],
                  "Strict-Transport-Security": ["max-age=31536000; includeSubDomains; preload"]
                }
              }
            },
            {
              "handler": "file_server",
              "root": "/var/www/html"
            }]
          }],
          "tls_connection_policies": [{
            "match": {
              "sni": ["xx.yy"]
            }
          }]
        }
      }
    },
    "tls": {
      "certificates": {
        "automate": ["xx.yy"]
      },
      "automation": {
        "policies": [{
          "issuers": [{
            "module": "acme",
            "email": "your@mail.com"
          },
          {
            "module": "acme",
            "ca": "https://acme.zerossl.com/v2/DV90",
            "email": "your@mail.com"
          }]
        }]
      }
    }
  }
}
```

**Comment by erroltuparker** [comment]
@WeidiDeng as im using unraid its basically this run command (cleaned up some of the extra bits)
```
docker run
  -d
  --name='dns'
  --net='bond0'
  --ip='192.168.86.1'
  --cpuset-cpus='16,17,18,19,20,21,22,23'
  --pids-limit 2048
  -v '/mnt/user/appdata/technitium-dnsserver/config':'/etc/dns/config':'rw'
  -v '/mnt/user/appdata/technitium-dnsserver/':'/etc/dns/':'rw'
  --hostname=dns
  --sysctl net.ipv6.conf.all.disable_ipv6=0
  --sysctl net.ipv6.conf.eth0.use_tempaddr=2
  --restart=unless-stopped 'technitium/dns-server'
```

As its on an actual network, all the ports are available to caddy.

Then I have Technitium configured as below 

![image](https://github.com/user-attachments/assets/22e38699-afa6-485c-8386-0027232cec28)


Then the caddyfile as configured below

```
	layer4 {
		tcp/:53 {
			@tdns {
				dns
			}
			route @tdns {
				proxy {
					upstream tcp/dns:538
					proxy_protocol v2
				}
			}
		}

		udp/:53 {
			@udns {
				dns
			}
			route @udns {
				proxy {
					proxy_protocol v2
					upstream udp/dns:538
				}
			}
		}

		udp/:853 {
			@quic {
				quic sni dns.example.com
			}
			route @quic {
				proxy {
					proxy_protocol v2
					upstream udp/dns:538
				}
			}
		}
		tcp/:853 {
			@dot {
				tls sni dns.example.com
			}
			route @dot {
				tls
				proxy {
					upstream tcp/dns:538
					proxy_protocol v2
				}
			}
		}
	}
```

**Comment by WeidiDeng** [comment]
@erroltuparker can you try `xcaddy build --with github.com/mholt/caddy-l4=github.com/WeidiDeng/caddy-l4@udp-pp` again to see if it's fixed?

**Comment by erroltuparker** [comment]
@WeidiDeng still the same issue but please see the results between proxy_protocol v1 and proxy_protocol v2

note: 192.168.86.5 is caddy

proxy_protocol v2 
```
[2025-01-14 08:23:37 UTC] [192.168.86.5:36809] [UDPPROXY] System.IO.InvalidDataException: The stream does not contain PROXY protocol header.
   at TechnitiumLibrary.Net.ProxyProtocol.ProxyProtocolStream.CreateAsServerAsync(Stream baseStream, CancellationToken cancellationToken) in Z:\Technitium\Projects\TechnitiumLibrary\TechnitiumLibrary.Net\ProxyProtocol\ProxyProtocolStream.cs:line 138
   at DnsServerCore.Dns.DnsServer.ReadUdpRequestAsync(Socket udpListener, DnsTransportProtocol protocol) in Z:\Technitium\Projects\DnsServer\DnsServerCore\Dns\DnsServer.cs:line 415
[2025-01-14 08:23:37 UTC] [192.168.86.5:36809] [UDPPROXY] System.ArgumentNullException: Value cannot be null. (Parameter 'address')
   at System.ArgumentNullException.Throw(String paramName)
   at System.Net.IPEndPoint..ctor(IPAddress address, Int32 port)
   at DnsServerCore.Dns.DnsServer.ReadUdpRequestAsync(Socket udpListener, DnsTransportProtocol protocol) in Z:\Technitium\Projects\DnsServer\DnsServerCore\Dns\DnsServer.cs:line 416
```

proxy_protocol v1
```
[2025-01-14 08:29:59 UTC] [192.168.86.5:59897] [UDPPROXY] System.IO.InvalidDataException: The stream does not contain PROXY protocol header.
   at TechnitiumLibrary.Net.ProxyProtocol.ProxyProtocolStream.CreateAsServerAsync(Stream baseStream, CancellationToken cancellationToken) in Z:\Technitium\Projects\TechnitiumLibrary\TechnitiumLibrary.Net\ProxyProtocol\ProxyProtocolStream.cs:line 138
   at DnsServerCore.Dns.DnsServer.ReadUdpRequestAsync(Socket udpListener, DnsTransportProtocol protocol) in Z:\Technitium\Projects\DnsServer\DnsServerCore\Dns\DnsServer.cs:line 415
[2025-01-14 08:29:59 UTC] [192.168.86.5:59897] [UDPPROXY] System.FormatException: An invalid IP address was specified.
 ---> System.Net.Sockets.SocketException (22): Invalid argument
   --- End of inner exception stack trace ---
   at TechnitiumLibrary.Net.ProxyProtocol.ProxyProtocolStream.ParseVersion1(String value) in Z:\Technitium\Projects\TechnitiumLibrary\TechnitiumLibrary.Net\ProxyProtocol\ProxyProtocolStream.cs:line 245
   at TechnitiumLibrary.Net.ProxyProtocol.ProxyProtocolStream.CreateAsServerAsync(Stream baseStream, CancellationToken cancellationToken)
   at DnsServerCore.Dns.DnsServer.ReadUdpRequestAsync(Socket udpListener, DnsTransportProtocol protocol) in Z:\Technitium\Projects\DnsServer\DnsServerCore\Dns\DnsServer.cs:line 415
```

**Comment by WeidiDeng** [comment]
proxy_protocol v1 can not be fixed, both upstream library doesn't support this use case. Only tcp is supported.

I think nginx can extract the exact address using this type of [syscall](https://stackoverflow.com/questions/5281409/get-destination-address-of-a-received-udp-packet). This requires rewriting packet connection handling to get the exact local address. For golang this local address is `0.0.0.0`, which may be rejected by DnsServer.

https://pkg.go.dev/golang.org/x/net@v0.34.0/ipv4#PacketConn.ReadFrom

https://pkg.go.dev/golang.org/x/net@v0.34.0/ipv6#PacketConn.ReadFrom

I won't have time to work on this for now.

**Comment by WeidiDeng** [comment]
@erroltuparker can you try `xcaddy build --with github.com/mholt/caddy-l4=github.com/WeidiDeng/caddy-l4@udp-pp` again to see if it's fixed?

No need to test v1 since it isn't supported by upstream at all.

**Comment by erroltuparker** [comment]
Hi @WeidiDeng, still same errors.

(btw, thanks for looking into this)

**Comment by lxhao61** [comment]
@skedastically @WeidiDeng 
But the program crashes when using the following example.
```
{
	log {
		level debug
	}
	layer4 {
		udp/:443 {
			@qsni quic sni xx.yy
			route @qsni {
				proxy {
					upstream udp/127.0.0.1:2443
				}
			}
			route {
				proxy {
					#proxy_protocol v2
					upstream udp/127.0.0.1:8443
				}
			}
		}
	}
}
```

**Comment by WeidiDeng** [comment]
@lxhao61 It should be fixed now.

**Comment by lxhao61** [comment]
@WeidiDeng 
The aforementioned BUG has been fixed, but another obvious BUG and an uncertain BUG were discovered through the following example tests.
```
{
	log {
		level debug
	}
	layer4 {
		tcp/:443 {
			@tsni tls sni zz.yy
			route @tsni {
				proxy {
					proxy_protocol v2
					upstream tcp/127.0.0.1:7443
				}
			}
			route {
				proxy {
					proxy_protocol v2
					upstream tcp/127.0.0.1:8443
				}
			}
		}
		udp/:443 {
			@qsni quic sni xx.yy
			route @qsni {
				proxy {
					upstream udp/127.0.0.1:2443
				}
			}
			route {
				proxy {
					#proxy_protocol v2
					upstream udp/127.0.0.1:8443
				}
			}
		}
	}
	servers 127.0.0.1:8443 {
		listener_wrappers {
			proxy_protocol {
				allow 127.0.0.1/32
			}
			tls
		}
	}
}
:8443, xx.zz:8443 {
	bind 127.0.0.1
	file_server {
		root /var/www/html
	}
}
```
BUG：
1. For ``tcp/127.0.0.1:7443``, proxy protocol transmission cannot be enabled. Everything works normally when both the sender and receiver have proxy protocol disabled.
2. For ``udp/127.0.0.1:8443``, proxy protocol transmission cannot be enabled. It is uncertain whether this is due to Caddy's HTTP/3 not supporting proxy protocol reception.

**Comment by lxhao61** [comment]
@WeidiDeng Hi, there’s a bug when forwarding with UDS. Master Xi, please fix it when you have time.

**Comment by erroltuparker** [comment]
@WeidiDeng any idea when this will be merged into the main?

It resolved my problem, but i need to make use of the new fail2ban option more currently so my udp proxying is no longer working until this has been merged.

Otherwise, do you think you can update your fork to include the latest changes?

**Comment by mholt** [comment]
Ah dang, GitHub didn't show me that comment (I guess it was hidden). Thanks for letting me know.

**Comment by lxhao61** [comment]
> Ah dang, GitHub didn't show me that comment (I guess it was hidden). Thanks for letting me know.

The content is too dense; it's easy to overlook details without a careful review. You're too kind. I hope this can help expedite the resolution of this issue.​

**Comment by polarathene** [comment]
Worth noting [are plans to align](https://github.com/mholt/caddy-l4/issues/176) the PROXY protocol lib [used in Caddy L4](https://github.com/mastercactapus/proxyprotocol) (_which has not been maintained for over 2 years_) to the one [used in the main Caddy](https://github.com/pires/go-proxyproto) project, but as has been mentioned earlier in this issue, [it lacks UDP support](https://github.com/mholt/caddy-l4/issues/269#issuecomment-2588866269) (_there does seem to be a [related feature request](https://github.com/pires/go-proxyproto/issues/88)? While [this comment](https://github.com/pires/go-proxyproto/issues/79#issuecomment-904588091) suggests some form of UDP support exists_).

That [linked comment from WeidiDeng](https://github.com/mholt/caddy-l4/issues/269#issuecomment-2588866269) also notes porting over to `pires/go-proxyproto` would be a big refactor, so migration might be a while off 😓 

> update: the new library doesn't support proxyprotocol for network other than tcp either.

Unclear if this was referring to library support in general, or just PROXY protocol v1 [which doesn't support UDP, thus no library will](https://github.com/pires/go-proxyproto/issues/45#issuecomment-716141685), [v2 is required for UDP support](https://github.com/traefik/traefik/issues/8613#issuecomment-2295918971).

**Comment by WeidiDeng** [comment]
Apparently, a `UnixConn` also implements `net.PacketConn` even if it's in stream mode. Should be easy to fix.

**Comment by WeidiDeng** [comment]
@lxhao61 can you try xcaddy build --with github.com/mholt/caddy-l4=github.com/WeidiDeng/caddy-l4@udp-unix to see if it's fixed?

**Comment by WeidiDeng** [comment]
In [another patch](https://github.com/mholt/caddy-l4/pull/307) I implemented local address for udp sockets on linux and windows sockets. It can be tested by `xcaddy build --with github.com/mholt/caddy-l4=github.com/WeidiDeng/caddy-l4@udp-local-addr`.

Feedbacks will be appreciated

**Comment by polarathene** [comment]
@lxhao61 they had a copy/paste typo in the command, the branch exists on their fork: https://github.com/WeidiDeng/caddy-l4/tree/udp-local-addr

```bash
xcaddy build --with github.com/WeidiDeng/caddy-l4@udp-local-addr
```

**Comment by lxhao61** [comment]
@mholt @WeidiDeng The pp-unix branch has fixed the UDS issue. Can it be merged?​

**Comment by WeidiDeng** [comment]
@lxhao61 udp-local-addr branch is supposed to merged after pp-unix branch.

**Comment by lxhao61** [comment]
> @mholt @WeidiDeng The pp-unix branch has fixed the UDS issue. Can it be merged?​

**Comment by lxhao61** [comment]
@WeidiDeng Oh, I see.​
